### ISSUE TITLE
How to get multiple values in params

### ISSUE THREAD

**Issue by azzamsa** [fukamachi/caveman#97]
@fukamachi Hi, Fukamachi.

I am sorry for disturbing your time and calmness.

I have tried this for more than  3 days, but I still can't solve this problem. I Have asked [this qestion](https://stackoverflow.com/questions/47324628/how-to-get-multiple-query-parameters-in-caveman) in Stackoverflow, but also no workaround.

I have tried this [caveman1 readme](https://github.com/fukamachi/caveman/blob/master/README.v1.markdown#multiple-values-in-params) but no luck.

thank you very much.

**Comment by azzamsa** [comment]
Stil there is no solution for this problem, but my dirty hack is: 

- add square brackets to the form name 
``` html

        <form action="/cbdua">
            <input type="checkbox" name="items[][name]" value="bike"> I have a bike<br>
            <input type="checkbox" name="items[][name]" value="car"> I have a car<br>
            <input type="submit" value="Checkbox 2!">
        </form>
```

- use  `_parsed` as `&key` then manipulating the result.

``` lisp
@route GET "/cbdua"
(lambda (&key _parsed)
  (let ((result (remove "name" (alexandria:flatten 
                               (loop for (vehicle . value)
                                  in _parsed collect value)) :test #'equal)))
    
    (let ((answer
           (cond
             ((equal hasil '(car)) "you have a car")
             ((equal hasil '(bike)) "you have a bike")
             ((equal hasil '(bike car))  "you have a car and bike")) )) 

      (render #P"cbdua.html"
              `(:answer ,answer
                         :result ,result)))))

```

**Comment by azzamsa** [comment]
This is my complete code to yield the result

``` lisp

(defvar *answers* nil)

(defun parse-params (input)
  (remove "name" (alexandria:flatten 
                  (loop for (vehicle . value)
                     in input collect value)) :test #'equal))

(defun get-result (input)
  (cond
    ((equal input '("car")) (setf *answers* "kamu punya car"))
    ((equal input '("bike")) (setf *answers* "kamu punya bike"))
    ((equal input '("bike" "car")) (setf *answers* "kamu punya car dan bike"))))
```

``` lisp
@route GET "/cbdua"
(defun cari-jawaban (&key _parsed)
  (render #P"cbdua.html" `(:jawaban ,(get-result (parse-params _parsed)))))
```

``` html
       <form action="/cbdua">
            <input type="checkbox" name="items[][name]" value="bike"> I have a bike<br>
            <input type="checkbox" name="items[][name]" value="car"> I have a car<br>
            <input type="submit" value="Checkbox 2!">
        </form>
```

**Comment by aarvid** [comment]
This is value of `_parsed`

`(("items" (("name" . "bike")) (("name" . "car"))))`

to get the value of parameter `items` do

`(cdr (assoc "items" _parsed :test #'string=))`

to get result

`((("name" . "bike")) (("name" . "car")))`

which is a list of associated lists.

This is because you html asks for a *matrix* of values `items[][name]`

if you change the html to `items[]`, you will get an *array* of values.

with both checked, `_parsed` will be:

`(("items" "bike" "car"))`

now

`(cdr (assoc "items" _parsed :test #'string=))`

returns

`("bike" "car")`

**Comment by azzamsa** [comment]
Hi, Aarvid.

It's long story to tell, I spent two days to find out the answer. I thought I was have wrong code, because the [caveman1 readme](https://github.com/fukamachi/caveman/blob/master/README.v1.markdown) tell to use `getf-all` for multiple value in params. I try to play with the result and no luck.

So, I think I have to use the "dirty hack" because in Caveman2 Readme, it does not tell how to get value in multiple params (eg. ?foo=1&foo=2). so instead of using `&key` I try to use `_parsed` and **manipulate** the result.

I play with `_parsed` more than 7 hours (as I am lisp newbie) to extract the result as it is associated list, so what I do is: 1. loop over items, get the value instead of key, and 2. remove the car value. so the result was ("bike" "car")

Even tough it was tedious, it's worth and it works :).

![2017-11-19-10 54 45 - dr kucing](https://user-images.githubusercontent.com/17734314/32992549-8b3047f4-cd7e-11e7-8f00-2d23b6972b57.png)



Unfortunately, I only focus to `items[][name]` and I though caveman does not supposed to use `items[]`

Thanks for pointing out.

I hope we will have good docs on caveman soon :)

---

Do you use caveman2 on your production?

**Comment by aarvid** [comment]
I have used caveman2 on a bunch toy/proof of concept systems and on one production system which is proprietary which is your basic data entry and data display site with some peculiar requirements but accessible only to my client and their clients.  So I can't show you the code or the website. I used caveman2 because it combines djula templates (my client was supposed to provide a css web designer but never did), a postgres database, internationalization (the site is in 3 languages and would be easy to add more. in fact I contributed to the Djula so it can generate cl-locale files and created [translate_client](https://github.com/aarvid/translate-client) to get google translations. )
